src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -1008,9 +1010,26 @@ private TokenValidationResult ValidateToken(string token, JsonWebToken outerToke
                 }
                 catch (Exception ex)
                 {
-                    // Keep going with the validation as the TokenValidationParameters may have the issuer and signing key set
+                    // The exception is not re-thrown here as the TokenValidationParameters may have the issuer and signing key set
                     // directly on them.
                     LogHelper.LogInformation(LogHelper.FormatInvariant(TokenLogMessages.IDX10261, validationParameters.ConfigurationManager.MetadataAddress, ex.ToString()));
+
+                    // Retry configuration retrieval on network failure (ServiceUnavailable or RequestTimeout).
+                    if (ex.GetType().Equals(typeof(InvalidOperationException)) && ex.Message.Contains("IDX20803 :")
+                        && ex.InnerException != null && ex.InnerException is IOException ioException
+                        && ioException.Data.Contains(HttpResponseConstants.StatusCode) &&
+                        (ioException.Data[HttpResponseConstants.StatusCode].Equals(HttpStatusCode.RequestTimeout) || ioException.Data[HttpResponseConstants.StatusCode].Equals(HttpStatusCode.ServiceUnavailable)))
+                    {
+                        try
+                        {
+                            LogHelper.LogInformation(LogHelper.FormatInvariant(TokenLogMessages.IDX10265, ioException.Data[HttpResponseConstants.StatusCode], ioException.Data[HttpResponseConstants.ResponseContent], validationParameters.ConfigurationManager.MetadataAddress));
+                            currentConfiguration = validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
+                        }
+                        catch (Exception retryException)
+                        {
+                            LogHelper.LogInformation(LogHelper.FormatInvariant(TokenLogMessages.IDX10261, validationParameters.ConfigurationManager.MetadataAddress, retryException.ToString()));
+                        }
+                    }
                 }
```

Comment:
I am not a fan of retrying in catch :).
Consider the following alternative using the catch/when:

                int maxAttempt = 2;
                for (int i = 1; i <= maxAttempt; i++)
                {
                    try
                    {
                        currentConfiguration = validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
                    }
                    catch (InvalidOperationException e) when
                            (e.Message.Contains("IDX20803 :") &&
                             e.InnerException != null && e.InnerException is IOException ioException &&
                             ioException.Data.Contains(HttpResponseConstants.StatusCode) &&
                             (ioException.Data[HttpResponseConstants.StatusCode].Equals(HttpStatusCode.RequestTimeout) || ioException.Data[HttpResponseConstants.StatusCode].Equals(HttpStatusCode.ServiceUnavailable)))
                    {
                        // The exception is not re-thrown here as the TokenValidationParameters may have the issuer and signing key set
                        // directly on them.
                        LogHelper.LogInformation(LogHelper.FormatInvariant(TokenLogMessages.IDX10261, validationParameters.ConfigurationManager.MetadataAddress, e.ToString()));

                        if (i < maxAttempt)
                            LogHelper.LogInformation(LogHelper.FormatInvariant(TokenLogMessages.IDX10265, ioException.Data[HttpResponseConstants.StatusCode], ioException.Data[HttpResponseConstants.ResponseContent], validationParameters.ConfigurationManager.MetadataAddress));
                    }
                    catch (Exception ex)
                    {
                        // The exception is not re-thrown here as the TokenValidationParameters may have the issuer and signing key set
                        // directly on them.
                        LogHelper.LogInformation(LogHelper.FormatInvariant(TokenLogMessages.IDX10261, validationParameters.ConfigurationManager.MetadataAddress, ex.ToString()));
                        break;
                    }
                }

Reply:
Thanks for this suggestion Danny! It definitely simplifies the code quite a bit. For ease of review (and my own sanity when rebasing 😅) I've combined this PR with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1785, which improves logging during this whole configuration retrieval process.

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1784/commits/47e0c5bf3236c4fc08d1f0f56dd250d5e6051ddb